duckduckgo/Android
# Adds tabs and data persistence

Task/Issue URLs: 
Tabs: https://app.asana.com/0/488551667048375/471193197990031
Implementation Task: https://app.asana.com/0/488551667048375/572068181812776
Persistence: https://app.asana.com/0/488551667048375/529727022544679

Tech Design URL: https://app.asana.com/0/481882893211075/543574221254520

**Description**:
Adds tabs and data persistance across launches

**Steps to test this PR**:

**TEST: Open tab**
1. Open a new tab from the browser
2. Open tab switcher, note that there are now two tabs in there
3. Switch between them

**TEST: Open tab from switcher**
1. Got to tab switcher
2. Open a new tab
3. Ensure new tab opens 
4. Open the tab switcher again and ensure old tab still exists and can be selected

**TEST: Long press to open a tab**
1. From the browser, long press on a url
2. Select "Open in new tab"
3. Ensure this opens in a new tab

**TEST: Open a tab from bookmarks**
1. Save a bookmark
2. Go to bookmarks and tap on saved bookmark
3. Ensure this opens in a new tab

**TEST: Closing tabs**
1. Got to tab switcher
2. Close tab
3. Go to browser again
4. Open switcher and ensure tab is gone

**TEST: Fire**
1. Press the browser fire button
2. Select "Clear all tabs and data"
3. Ensure all tabs are closed
4. Repeat the same from the switcher

**TEST: Persistence**
1. Open some new tabs
2. Kill the app
3. Relaunch the app and ensure that it opens on the tab you were last on

**TEST: Launch modes**
Perform the following
• Share link/text from external app to our app
• Open the app via the widget
• Open the app via the assistant

For each of the following states:
• App closed
• App running and on home page
• App running and on browser page

**TEST: Database upgrade**
1. Install the previously released version of the app (a build from master should do)
1. Save some bookmarks
1. Then install this version over the top
1. Ensure the bookmarks are still ok

**TEST: And anything else you can think of!**

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
